### PR TITLE
scalpel: init at 2.1

### PR DESCRIPTION
**Pull request by shard77**
## Description of changes
Scalpel - A file carving and indexing application. It recover files based on their headers, footers and internal data structures, the tool is based on Foremost.

See:
- https://repology.org/project/scalpel/information
- https://github.com/sleuthkit/scalpel

Added the backport label, since this is a new package.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:5]:
```suggestion
, autoreconfHook
```

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:21]:
```suggestion
  nativeBuildInputs = [ 
    autoreconfHook 
    libtool 
    tre 
  ];
```

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:29]: Shouldn't this be the default?

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:37]:
```suggestion
```

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:33]: the autoreconfHook does this already

```suggestion
```

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:51]: 1. Which binary files it does include?
2. include `fromSource` as well

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:52]:
```suggestion
    license = with licenses; [ asl20 ];
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Should we include an explicit abandonware into Nixpkgs?

**Comment by shard77** [comment]
> Should we include an explicit abandonware into Nixpkgs?

Despite being abandoned, it is still employed in the field of DFIR/Forensics. Such projects, often found in cybersecurity, remain in use even if discontinued because they function well and are beneficial in certain cases.

File: pkgs/by-name/sc/scalpel/package.nix
Comment (shard77) [reply pkgs/by-name/sc/scalpel/package.nix:51]: Actually `binaryByteCode` can be removed, I was checking the usage of certain libraries in the code, but the only lib is `libscalpel` and it's built from source.

File: pkgs/by-name/sc/scalpel/package.nix
Comment (shard77) [reply pkgs/by-name/sc/scalpel/package.nix:29]: Without it, it was using C++17.

File: pkgs/by-name/sc/scalpel/package.nix
Comment (shard77) [reply pkgs/by-name/sc/scalpel/package.nix]: Tried using `autoreconfHook` , I get some runtime errors on build, and would need to create a patch for `Makefile.am`. Would it be better to rollback and use `autoconf` and `automake` or to create the patch file and use `autoreconfHook` ?

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:32]: Use this instead:

```suggestion
  env.NIX_CFLAGS_COMPILE = toString [ "-std=c++14" ];
```

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:36]: Remove this line

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:41]: More elegant:

```suggestion
    install -Dm644 scalpel.conf -t $out/share/scalpel/
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by shard77** [comment]
@AndersonTorres applied your changes, thank you. I am currently trying to make a patch since `autoreconfHook` throws me some errors:
```
Makefile.am: installing './depcomp'
Makefile.am:3: warning: 'LDFLAGS' is a user variable, you should not override it;
Makefile.am:3: use 'AM_LDFLAGS' instead
src/Makefile.am:3: warning: 'LDFLAGS' is a user variable, you should not override it;
src/Makefile.am:3: use 'AM_LDFLAGS' instead
autoreconf: error: automake failed with exit status: 1
```
In this case, would it be better to rollback and use `autoconf` and `automake` instead of `autoreconfHook`, or just provide a patch? Still trying to localize the issue, it doesn't seem like the warnings are the source of the error. And also, if I provide a patch for the `Makefile.am`, should I proceed with using `patches = [ ];` or just use text substitution in `postPatch` ?

**Comment by AndersonTorres** [comment]
This looks too complicated to a mere packager tries to fix.

Let's pick your original idea of raw autoconf and automake.

**Comment by shard77** [comment]
Fixed it, builds as expected.

File: pkgs/by-name/sc/scalpel/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/sc/scalpel/package.nix:30]: Just to know:
This config file can be overridden?

Comment (shard77) [reply pkgs/by-name/sc/scalpel/package.nix:30]: Yes, basically there is a default user configuration that you can override. And you need to specify a default one.

Comment (AndersonTorres) [reply pkgs/by-name/sc/scalpel/package.nix:30]: This is fine.

Maybe the only concern my paranoid side would point out is if this default config file is not referring to dangerous paths, or if any syntax errors on this file would prevent the whole program to run.

**Review by AndersonTorres — APPROVED** [review]

**Review by D3vil0p3r — APPROVED** [review]
LGTM

File: pkgs/by-name/sc/scalpel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sc/scalpel/package.nix:17]:
```suggestion
    rev = "35e1367ef2232c0f4883c92ec2839273c821dd39";
```
full hash required -- repo could go active and get millions of commits tmw, never know.

File: pkgs/by-name/sc/scalpel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sc/scalpel/package.nix:34]:
```suggestion
```

File: pkgs/by-name/sc/scalpel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sc/scalpel/package.nix:41]:
```suggestion
    "--with-pic"
    "CXXFLAGS=-std=c++14"
```

**Review by ghost — COMMENTED** [review]
looks good. just a few nits.

**Comment by shard77** [comment]
> looks good. just a few nits.

Yep, those changes make sense. Thanks!

**Review by ghost — COMMENTED** [review]
missed that the darwin build is failing. need to add `CXXFLAGS=-std=c++14 -Wno-error=reserved-user-defined-literal` but the `configureFlags` doesn't allow spaces, but the env var does so i guess back to

`env.CXXFLAGS = "-std=c++14 -Wno-error=reserved-user-defined-literal";`


which will build on x64 darwin. sorry about that.
<details>
<summary>x64 darwin build is failing</summary>
<pre>
dig.cpp:381:58: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
        fprintf(state->auditFile, "\nSkipped the first %"PRIu64 "bytes of %s...\n",
                                                         ^

dig.cpp:384:52: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
            fprintf(stdout, "\nSkipped the first %"PRIu64 "bytes of %s...\n",
                                                   ^

dig.cpp:621:63: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                fprintf(stdout, "A %s header was found at : %"PRIu64 "\n",
                                                              ^

dig.cpp:650:86: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                            "Memory reallocation performed, total header storage = %"PRIu64 "\n",
                                                                                     ^
                                                                                
dig.cpp:759:63: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                fprintf(stdout, "A %s footer was found at : %"PRIu64 "\n",
                                                              ^

dig.cpp:787:86: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                            "Memory reallocation performed, total footer storage = %"PRIu64 "\n",
                                                                                     ^
                                                                                
dig.cpp:994:41: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                fprintf(stdout, "Read %"PRIu64 " bytes from image file.\n", bytesread);
                                        ^

dig.cpp:1109:47: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
        fprintf(stdout, "Total file size is %"PRIu64 " bytes\n", filesize);
                                              ^

dig.cpp:1540:35: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
        fprintf(stdout, "\" --> %"PRIu64 " files\n", currentneedle->numfilestocarve);
                                  ^

dig.cpp:1918:64: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
        fprintf(stdout, "# of blocks in coverage blockmap is %"PRIu64 ".\n",
                                                               ^

dig.cpp:1973:64: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
        fprintf(stdout, "# of blocks in coverage blockmap is %"PRIu64 ".\n",
                                                               ^

dig.cpp:2287:80: warning: unused parameter 'inReader' [-Wunused-parameter]
fseeko_use_coverage_map(struct scalpelState *state, ScalpelInputReader * const inReader, off64_t offset) {
                                                                               ^
dig.cpp:2305:27: warning: comparison of integers of different signs: 'unsigned long long' and 'off_t' (aka 'long long') [-Wsign-compare]
        while (totalbytes < (offset > 0 ? offset : offset * -1) &&
               ~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dig.cpp:2327:35: warning: comparison of integers of different signs: 'unsigned long long' and 'off_t' (aka 'long long') [-Wsign-compare]
                    && totalbytes < (offset > 0 ? offset : offset * -1)) {
                       ~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
dig.cpp:2377:68: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                "Coverage map decreased current file position by %"PRIu64 " bytes.\n",
                                                                   ^

dig.cpp:2353:95: warning: unused parameter 'inReader' [-Wunused-parameter]
static off64_t ftello_use_coverage_map(struct scalpelState *state, ScalpelInputReader * const inReader) {
                                                                                              ^
dig.cpp:2402:59: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                "Issuing coverage map-based READ, wants %"PRIu64 " bytes.\n",
                                                          ^

dig.cpp:2428:61: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                        "fread using coverage map to skip %"PRIu64 " bytes.\n", bytestoskip);
                                                            ^

dig.cpp:2453:59: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                        "fread using coverage map found %"PRIu64 " consecutive bytes.\n",
                                                          ^

dig.cpp:2468:70: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                    fprintf(stdout, "fread using coverage map read %"PRIu64 " bytes.\n",
                                                                     ^

dig.cpp:2882:35: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
            if(fprintf(dbfile, "%"PRIu64 "\n", currentneedle->offsets.numheaders)
                                  ^

dig.cpp:2916:35: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
            if(fprintf(dbfile, "%"PRIu64 "\n", currentneedle->offsets.numfooters)
                                  ^

dig.cpp:2928:33: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                    (dbfile, "%"PRIu64 "\n",
                                ^

5 warnings and 19 errors generated.
make[1]: *** [Makefile:536: dig.lo] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-scalpel-2.1.drv-0/source/src'
make: *** [Makefile:534: all-recursive] Error 1
</pre>
</details>

File: pkgs/by-name/sc/scalpel/package.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/sc/scalpel/package.nix]: I think `sourceProvenance` can be removed -- can you point me to where the binary is that is distributed? i did a quick look and didn't find one, unless it is embedded in the code or the archive...

**Comment by shard77** [comment]
> missed that the darwin build is failing. need to add `CXXFLAGS=-std=c++14 -Wno-error=reserved-user-defined-literal` but the `configureFlags` doesn't allow spaces, but the env var does so i guess back to
> 
> `env.CXXFLAGS = "-std=c++14 -Wno-error=reserved-user-defined-literal";`
> 
> which will build on x64 darwin. sorry about that.
> x64 darwin build is failing

@a-n-n-a-l-e-e No worries, should be good now. Thanks.

File: pkgs/by-name/sc/scalpel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sc/scalpel/package.nix:42]:
```suggestion
```

File: pkgs/by-name/sc/scalpel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sc/scalpel/package.nix:55]:
```suggestion
```
when looking at result i do not see any binaries distributed -- looking at the source there are some dlls for windows but that's it. could you point out where the binaryNativeCode is?

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/sc/scalpel/package.nix:54]: eg: everything that ends up in the nix store is compiled from source. which doesn't count as binaryNativeCode, I believe that `sourceProvenance` binaryNatvieCode is used when the just re-distributing compiled code and not building it. so I think that this line can be removed -- unless i am missing something.

Comment (shard77) [reply pkgs/by-name/sc/scalpel/package.nix:54]: @a-n-n-a-l-e-e I thought [some of these files](https://github.com/sleuthkit/scalpel/tree/master/pthreads-win32) were needed as libs (especially [libpthreadGC2.a](https://github.com/sleuthkit/scalpel/blob/master/pthreads-win32/libpthreadGC2.a) and [libpthreadGCE2.a](https://github.com/sleuthkit/scalpel/blob/master/pthreads-win32/libpthreadGCE2.a). I'm going to check again.

Also, is it normal, that when I try building with the updated changes for the derivation, I get:
`cc1plus: error: '-Wno-error=reserved-user-defined-literal': no option '-Wreserved-user-defined-literal'` ?

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/sc/scalpel/package.nix:54]: > @a-n-n-a-l-e-e I thought [some of these files](https://github.com/sleuthkit/scalpel/tree/master/pthreads-win32) were needed as libs (especially [libpthreadGC2.a](https://github.com/sleuthkit/scalpel/blob/master/pthreads-win32/libpthreadGC2.a) and [libpthreadGCE2.a](https://github.com/sleuthkit/scalpel/blob/master/pthreads-win32/libpthreadGCE2.a). I'm going to check again.

thanks. if that is the case it is loading the win32 blobs then i'd assume it is x86 only and aarch64 not supported.

> Also, is it normal, that when I try building with the updated changes for the derivation, I get: `cc1plus: error: '-Wno-error=reserved-user-defined-literal': no option '-Wreserved-user-defined-literal'` ?

argh. i guess that option is clang only. so need to add it conditionally.
```nix
env.CXXFLAGS = "-other flags" + lib.optionalString stdenv.cc.isClang "  -Wno-error=reserved-user-defined-literal";
```

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/sc/scalpel/package.nix:54]: > @a-n-n-a-l-e-e I thought [some of these files](https://github.com/sleuthkit/scalpel/tree/master/pthreads-win32) were needed as libs (especially [libpthreadGC2.a](https://github.com/sleuthkit/scalpel/blob/master/pthreads-win32/libpthreadGC2.a) and [libpthreadGCE2.a](https://github.com/sleuthkit/scalpel/blob/master/pthreads-win32/libpthreadGCE2.a). I'm going to check again.

these do not seem to be needed, i just tested by adding.
```nix
 preBuild = ''
    rm -f pthreads-win32/*.dll pthreads-win32/*.a
  '';
```
and everything built fine on darwin

Comment (shard77) [reply pkgs/by-name/sc/scalpel/package.nix:54]: @a-n-n-a-l-e-e I double checked, you are right, I'll remove `binaryNativeCode`. I re-built with the changes and it works. Thanks again.

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/sc/scalpel/package.nix]: > @a-n-n-a-l-e-e I double checked, you are right, I'll remove `binaryNativeCode`. I re-built with the changes and it works. Thanks again.

remove the line please.
https://nixos.org/manual/nixpkgs/stable/#sec-meta-sourceProvenance
when not set built from source is implied.

**Comment by shard77** [comment]
Seems like the first check is not going through for some reason.. I don't really understand the issue here.

File: pkgs/by-name/sc/scalpel/package.nix
Comment (D3vil0p3r) [reply pkgs/by-name/sc/scalpel/package.nix]: Why a check saying "package.nix" not existing is failing if it exists?

**Comment by a-n-n-a-l-e-e** [comment]
> Seems like the first check is not going through for some reason.. I don't really understand the issue here. The error `error: path '/home/runner/work/nixpkgs/nixpkgs/"pkgs/by-name/sc/scalpel/package.nix"' does not exist` stipulates that my `package.nix` file is not existing. Didn't do any changes except squashing the commits and editing two lines.

me neither but if you remove the `sourceProvinance` line and then push the change perhaps it will resolve.

File: pkgs/by-name/sc/scalpel/package.nix
Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/sc/scalpel/package.nix]: looks like a recent commit https://github.com/NixOS/nixpkgs/pull/279892 might be causing this and there is a PR for the fix: https://github.com/NixOS/nixpkgs/pull/281366

File: pkgs/by-name/sc/scalpel/package.nix
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sc/scalpel/package.nix:54]:
```suggestion
```
not needed as fromSource is the default.
https://nixos.org/manual/nixpkgs/stable/#sec-meta-sourceProvenance

**Review by ghost — CHANGES_REQUESTED** [review]
looks good please remove the one line -- and once the CI is fixed can merge.

Result of `nixpkgs-review pr 278750` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>scalpel</li>
  </ul>
</details>

---

Result of `nixpkgs-review pr 278750` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>scalpel</li>
  </ul>
</details>

**Comment by a-n-n-a-l-e-e** [comment]
@ofborg build scalpel

**Comment by a-n-n-a-l-e-e** [comment]
> Should be good to go now.

thanks! looks good. i will merge once the ofborg-eval and aarch64-darwin succeed -- which could take awhile

**Review by ghost — APPROVED** [review]

**Pull request merged**
